File: robitcubebot_controller/src/twist_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist_stamped.hpp"
#include "std_msgs/msg/header.hpp"

#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

class PublisherNode : public rclcpp::Node
{
public:
    PublisherNode() : Node("teleop_publisher")
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>("/robitcubebot_controller/cmd_vel", 10);
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            std::bind(&PublisherNode::publish_twist, this));
        RCLCPP_INFO(this->get_logger(), "Publisher node initialized");
        old_settings = get_keyboard_settings();
        set_keyboard_nonblocking();
    }

    ~PublisherNode()
    {
        reset_keyboard_settings();
    }

private:
    void publish_twist()
    {
        auto twist_msg = geometry_msgs::msg::TwistStamped();
        twist_msg.header = std_msgs::msg::Header();
        twist_msg.header.stamp = this->now();
        twist_msg.header.frame_id = "";

        twist_msg.twist = get_keyboard_input();

        publisher_->publish(twist_msg);
        RCLCPP_INFO(this->get_logger(), "Publishing TwistStamped message");
    }

    geometry_msgs::msg::Twist get_keyboard_input()
    {
        char key = 0;
        geometry_msgs::msg::Twist cmd_vel_msg;

        if (read(STDIN_FILENO, &key, 1) > 0)
        {
            switch (key)
            {
            case 'w':
                cmd_vel_msg.linear.x = 0.5;  // Move forward
                cmd_vel_msg.angular.z = 0.0;
                break;
            case 'x':
                cmd_vel_msg.linear.x = -0.5;  // Move backward
                cmd_vel_msg.angular.z = 0.0;
                break;
            case 'a':
                cmd_vel_msg.linear.x = 0.0;
                cmd_vel_msg.angular.z = 1.0;  // Turn left
                break;
            case 'd':
                cmd_vel_msg.linear.x = 0.0;
                cmd_vel_msg.angular.z = -1.0;  // Turn right
                break;
            case 's':
                cmd_vel_msg.linear.x = 0.0;
                cmd_vel_msg.angular.z = 0.0;  // Stop
                break;
            default:
                break;
            }
        }

        return cmd_vel_msg;
    }

    termios get_keyboard_settings()
    {
        termios settings;
        tcgetattr(STDIN_FILENO, &settings);
        return settings;
    }

    void set_keyboard_nonblocking()
    {
        termios new_settings = old_settings;
        new_settings.c_lflag &= ~(ICANON | ECHO);
        tcsetattr(STDIN_FILENO, TCSANOW, &new_settings);
        fcntl(STDIN_FILENO, F_SETFL, fcntl(STDIN_FILENO, F_GETFL) | O_NONBLOCK);
    }

    void reset_keyboard_settings()
    {
        tcsetattr(STDIN_FILENO, TCSANOW, &old_settings);
    }

    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    termios old_settings;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: robitcubebot_controller/src/robitcubebot_controller/twist_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped, Twist
from std_msgs.msg import Header
import sys, select, termios, tty

class PublisherNode(Node):
    def __init__(self):
        super().__init__('teleop_publisher')
        self.publisher_ = self.create_publisher(TwistStamped, '/robitcubebot_controller/cmd_vel', 10)
        self.timer_ = self.create_timer(0.1, self.publish_twist)
        self.get_logger().info('Publisher node initialized')
        self.old_settings = termios.tcgetattr(sys.stdin)
        tty.setcbreak(sys.stdin.fileno())

    def publish_twist(self):
        twist_msg = TwistStamped()
        twist_msg.header = Header()
        twist_msg.header.stamp = self.get_clock().now().to_msg()
        twist_msg.header.frame_id = ''

        twist_msg.twist = self.get_keyboard_input()

        self.publisher_.publish(twist_msg)
        self.get_logger().info('Publishing TwistStamped message')

    def get_keyboard_input(self):
        key = None
        if select.select([sys.stdin], [], [], 0) == ([sys.stdin], [], []):
            key = sys.stdin.read(1)
        cmd_vel_msg = Twist()
        if key == 'w':
            cmd_vel_msg.linear.x = 0.5  # Move forward
            cmd_vel_msg.angular.z = 0.0
        elif key == 'x':
            cmd_vel_msg.linear.x = -0.5  # Move backward
            cmd_vel_msg.angular.z = 0.0
        elif key == 'a':
            cmd_vel_msg.linear.x = 0.0
            cmd_vel_msg.angular.z = 1.0  # Turn left
        elif key == 'd':
            cmd_vel_msg.linear.x = 0.0
            cmd_vel_msg.angular.z = -1.0  # Turn right
        elif key == 's':  # Spacebar to stop
            cmd_vel_msg.linear.x = 0.0
            cmd_vel_msg.angular.z = 0.0
        return cmd_vel_msg

def main(args=None):
    rclpy.init(args=args)
    publisher_node = PublisherNode()
    try:
        rclpy.spin(publisher_node)
    except KeyboardInterrupt:
        pass
    finally:
        termios.tcsetattr(sys.stdin, termios.TCSADRAIN, publisher_node.old_settings)
        publisher_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
